pkg/sql/schema_adapter.go

```diff
@@ -74,69 +117,100 @@ func (s DefaultSchema) SchemaInitializingQueries(topic string) []string {
 }
 
-func (s DefaultSchema) InsertQuery(topic string, msgs message.Messages) (string, []interface{}, error) {
+func (s DefaultMySQLSchema) InsertQuery(topic string, msgs message.Messages) (string, []interface{}, error) {
 	insertQuery := fmt.Sprintf(
 		`INSERT INTO %s (uuid, payload, metadata) VALUES %s`,
 		s.MessagesTable(topic),
 		strings.TrimRight(strings.Repeat(`(?,?,?),`, len(msgs)), ","),
 	)
 
-	var args []interface{}
-	for _, msg := range msgs {
-		metadata, err := json.Marshal(msg.Metadata)
-		if err != nil {
-			return "", nil, errors.Wrapf(err, "could not marshal metadata into JSON for message %s", msg.UUID)
-		}
-
-		args = append(args, msg.UUID, msg.Payload, metadata)
+	args, err := defaultInsertArgs(msgs)
+	if err != nil {
+		return "", nil, err
 	}
 
 	return insertQuery, args, nil
 }
 
-func (s DefaultSchema) SelectQuery(topic string, consumerGroup string, offsetsAdapter OffsetsAdapter) (string, []interface{}) {
+func (s DefaultMySQLSchema) SelectQuery(topic string, consumerGroup string, offsetsAdapter OffsetsAdapter) (string, []interface{}) {
 	nextOffsetQuery, nextOffsetArgs := offsetsAdapter.NextOffsetQuery(topic, consumerGroup)
-
 	selectQuery := `
-		SELECT offset,uuid,payload,metadata FROM ` + s.MessagesTable(topic) + `
+		SELECT offset, uuid, payload, metadata FROM ` + s.MessagesTable(topic) + `
 		WHERE 
 			offset > (` + nextOffsetQuery + `)
 		ORDER BY 
-		 offset ASC 
+			offset ASC
 		LIMIT 1`
 
 	return selectQuery, nextOffsetArgs
 }
 
-type defaultSchemaRow struct {
-	Offset   int64
-	UUID     []byte
-	Payload  []byte
-	Metadata []byte
+func (s DefaultMySQLSchema) UnmarshalMessage(row *sql.Row) (offset int, msg *message.Message, err error) {
+	return unmarshalDefaultMessage(row)
 }
 
-func (s DefaultSchema) UnmarshalMessage(row *sql.Row) (offset int, msg *message.Message, err error) {
-	r := defaultSchemaRow{}
-	err = row.Scan(&r.Offset, &r.UUID, &r.Payload, &r.Metadata)
-	if err != nil {
-		return 0, nil, errors.Wrap(err, "could not scan message row")
+func (s DefaultMySQLSchema) MessagesTable(topic string) string {
+	if s.GenerateMessagesTableName != nil {
+		return s.GenerateMessagesTableName(topic)
 	}
+	return fmt.Sprintf("`watermill_%s`", topic)
+}
 
-	msg = message.NewMessage(string(r.UUID), r.Payload)
+// DefaultPostgresSchema is a default implementation of SchemaAdapter based on PostgreSQL.
+type DefaultPostgresSchema struct {
+	// GenerateMessagesTableName may be used to override how the messages table name is generated.
+	GenerateMessagesTableName func(topic string) string
+}
 
-	if r.Metadata != nil {
-		err = json.Unmarshal(r.Metadata, &msg.Metadata)
-		if err != nil {
-			return 0, nil, errors.Wrap(err, "could not unmarshal metadata as JSON")
-		}
+func (s DefaultPostgresSchema) SchemaInitializingQueries(topic string) []string {
+	createMessagesTable := strings.Join([]string{
+		`CREATE TABLE IF NOT EXISTS ` + s.MessagesTable(topic) + ` (`,
+		`"offset" SERIAL,`,
+		`"uuid" VARCHAR(36) NOT NULL,`,
```

Comment:
doesn't it have built-in uuid type?

Reply:
It does, although it's used for UUID only, so it's not that universal (won't accept an ULID for example).